lib/vanagon/driver.rb

```diff
@@ -129,18 +129,18 @@ def run # rubocop:disable Metrics/AbcSize
       @engine.dispatch("(cd #{@engine.remote_workdir}; #{@platform.make})")
       @engine.retrieve_built_artifact
 
-      unless @preserve
+      if %i[never on-failure].include? @preserve
         @engine.teardown
         cleanup_workdir
       end
     rescue => e
+      if [:never].include? @preserve
+        @engine.teardown
+        cleanup_workdir
+      end
       $stderr.puts e
       $stderr.puts e.backtrace.join("\n")
       raise e
-    ensure
-      if ["hardware", "ec2"].include?(@engine.name)
-        @engine.teardown
-      end
```

Comment:
I think we want hardware/ec2 platforms to never get preserved so we don't end up leaving these up either blocking other people or costing money.
@McdonaldSeanp might know more though

Reply:
I can't speak to the ec2 engine, but yes the hardware should always be torn down.